src/Proto3/Suite/Class.hs

```diff
@@ -485,51 +484,61 @@ instance MessageField BL.ByteString
 instance (Named e, ProtoEnum e) => MessageField (Enumerated e)
 
 instance (Ord k, Primitive k, MessageField k, Primitive v, MessageField v) => MessageField (M.Map k v) where
-  encodeMessageField num = foldMap (Encode.embedded num . encodeMessage (fieldNumber 1)) . M.toList
+  encodeMessageField num =
+    let encodeField :: k -> v -> Encode.MessageBuilder
+        encodeField k v = Encode.embedded num (encodeMessage 1 (k, v))
+     in M.foldrWithKey (\k -> (<>) . encodeField k) mempty
```

Comment:
> we would need to check the core generated by GHC

@j6carey Do you think it would be worth using [`inspection-testing`](https://github.com/nomeata/inspection-testing) in `proto3-wire` or `proto3-suite` to ensure optimizations take effect in Core?

Reply:
@evanrelf , that's an interesting idea, but probably should not block this PR, IMO.

Reply:
I agree, should not block this PR 👍 Just wanted to mention the idea in general.